File: tests/xcalablemp/global-view/loop/F/reduction_max.f90
program main
  include 'xmp_lib.h'
  integer,parameter:: N=10
  integer random_array(N), ans_val, result
!$xmp nodes p(*)
!$xmp template t1(N,N,N)
!$xmp template t2(N,N,N)
!$xmp template t3(N,N,N)
!$xmp distribute t1(block,*,*) onto p
!$xmp distribute t2(*,block,*) onto p
!$xmp distribute t3(*,*,block) onto p
  integer a(N), sa
  real*8  b(N), sb
  real*4  c(N), sc
  real tmp(N)
!$xmp align a(i) with t1(i,*,*)
!$xmp align b(i) with t2(*,i,*)
!$xmp align c(i) with t3(*,*,i)

  result = 0
  call random_number(tmp)
  random_array(:) = int(tmp(:) * 10000)

!$xmp loop (i) on t1(i,:,:)
  do i=1, N
     a(i) = random_array(i)
  enddo

!$xmp loop (i) on t2(:,i,:)
  do i=1, N
     b(i) = dble(random_array(i))
  enddo

!$xmp loop (i) on t3(:,:,i)
  do i=1, N
     c(i) = real(random_array(i))
  enddo

  ans_val = 0
  do i=1, N
     ans_val = max(ans_val, random_array(i))
  enddo

  sa = 0
  sb = 0.0
  sc = 0.0

!$xmp loop (i) on t1(i,:,:) reduction(max: sa)
  do i=1, N
     sa = max(sa, a(i))
  enddo

!$xmp loop (i) on t2(:,i,:) reduction(max: sb)
  do i=1, N
     sb = max(sb, b(i))
  enddo

!$xmp loop (i) on t3(:,:,i) reduction(max: sc)
  do i=1, N
     sc = max(sc, c(i))
  enddo
  
  if( sa .ne. ans_val .or. sb .ne. dble(ans_val) .or. sc .ne. real(ans_val) ) then
     result = -1
  endif

!$xmp reduction(+:result)
!$xmp task on p(1)
  if( result .eq. 0 ) then
     write(*,*) "PASS"
  else
     write(*,*) "ERROR"
     call exit(1)
  endif
!$xmp end task

end program main

File: tests/xcalablemp/global-view/loop/C-Round/reduction_max.c
#include <xmp.h>
#include <stdio.h>  
#include <stdlib.h> 
static const int N=1000;
int random_array[1000];
#pragma xmp nodes p(*)
#pragma xmp template t1(0:N-1,0:N-1,0:N-1)
#pragma xmp template t2(0:N-1,0:N-1,0:N-1)
#pragma xmp template t3(0:N-1,0:N-1,0:N-1)
#pragma xmp distribute t1(block,*,*) onto p
#pragma xmp distribute t2(*,block,*) onto p
#pragma xmp distribute t3(*,*,block) onto p
int a[N],sa=0;
double b[N],sb=0.0;
float c[N],sc=0.0;
int i,k,result=0,ans_val=0;
#pragma xmp align a[i] with t1(i,*,*)
#pragma xmp align b[i] with t2(*,i,*)
#pragma xmp align c[i] with t3(*,*,i)

int main(void)
{
  srand(0);
  for(i=0;i<N;i++)
    random_array[i] = rand();
    
#pragma xmp loop on t1(i,:,:)
  for(i=0;i<N;i++)
    a[i] = random_array[i];

#pragma xmp loop on t2(:,i,:)
  for(i=0;i<N;i++)
    b[i] = random_array[i];
  
#pragma xmp loop on t3(:,:,i)
  for(i=0;i<N;i++)
    c[i] = random_array[i];

  for(i=0;i<N;i++)
    if(random_array[i]>ans_val)
      ans_val=random_array[i];

#pragma xmp loop (i) on t1(i,:,:) reduction(max:sa)
  for(i=0;i<N;i++)
    if(a[i]>sa)
      sa=a[i];
  
#pragma xmp loop (i) on t2(:,i,:) reduction(max:sb)
  for(i=0;i<N;i++)
    if(b[i]>sb)
      sb=b[i];
  
#pragma xmp loop (i) on t3(:,:,i) reduction(max:sc)
  for(i=0;i<N;i++)
    if(c[i]>sc)
      sc=c[i];

  if( (sa != ans_val) || (sb != (double)ans_val) || (sc != (float)ans_val) )
    result = -1;


#pragma xmp reduction(+:result)
#pragma xmp task on p(1)
  {
    if(result == 0){
      printf("PASS\n");
    }
    else{
      fprintf(stderr, "ERROR\n");
      exit(1);
    }
  }

  return 0;
}
